### PR TITLE
Add throttling support for callbacks

### PR DESCRIPTION
**Pull request by srajko**

### CODE REVIEW

**Comment by johnhaley81** [comment]
This will fix #919

**Comment by johnhaley81** [comment]
@srajko can you add a test to confirm that the callback is indeed being throttled? It could be another clone test.

**Comment by srajko** [comment]
The changes are as follows:

- Progress callbacks (`git_checkout_progress_cb`, `git_diff_file_cb`, `git_stash_apply_progress_cb`, `git_transfer_progress_cb`) now have default throttling of ~1 callback per 100ms
- All callbacks provided through configuration objects can now be throttled - instead of passing a function, pass an object `{ callback, throttle }` where `throttle` is the throttle period (~1 callback per throttle period)

A rundown of the commits:
 - `Add AsyncBatonWithResult and ExecuteAsync` - just refactoring, eliminating duplicated code
 - `Store defaultResult in baton` - I originally tried doing the throttling in the baton but that ended up being a mistake.  This moved the default result to the baton so the baton could use it.  I don't think the change is necessary for the final version of this PR, but I still like moving things out of combyne-templated files so I kept it.
 - `Refactor structs to use CallbackWrapper for callbacks` - just refactoring, leverages the previously unused `CallbackWrapper` to set us up for bundling throttling state with the javascript callback.
 - `Add callback throttle functionality` - the actual throttle
 - `Specify throttle in progress callback descriptors` - default settings for progress callbacks.

**Comment by johnhaley81** [comment]
:+1: nice job! Thanks @srajko

File: generate/templates/manual/include/async_baton.h [hunk 1]
```diff
@@ -13,4 +16,33 @@ struct AsyncBaton {
   bool done;
 };
 
+template<typename ResultT>
```
Comment (johnhaley81) [inline generate/templates/manual/include/async_baton.h]: :100:

**Pull request merged**